### PR TITLE
Enable message filtering to reduce network traffic.

### PR DESCRIPTION
**Pull request by MatejFranceskin**
Add Filter field to endpoint configuration to limit the messages that are sent to particular endpoint and improve performance.
For example XYZ is interested only in messages 24,33,76,320,321,323.
This configuration filters all other messages so that they're not sent to XYZ and don't need to be parsed.

[UdpEndpoint XYZ]
Mode = Normal
Address = 127.0.0.1
Port = 14530
Filter = 24,33,76,320,321,323

### CODE REVIEW

File: src/mavlink-router/endpoint.cpp [hunk 1]
```diff
@@ -70,20 +70,20 @@ int Endpoint::handle_read()
 {
     int target_sysid, target_compid, r;
     uint8_t src_sysid, src_compid;
+    uint32_t msg_id;
     struct buffer buf{};
 
-    while ((r = read_msg(&buf, &target_sysid, &target_compid, &src_sysid, &src_compid)) > 0)
+    if ((r = read_msg(&buf, &target_sysid, &target_compid, &src_sysid, &src_compid, &msg_id)) > 0)
```
Comment (lucasdemarchi) [inline src/mavlink-router/endpoint.cpp]: should continue to be `while`, or put in a separate commit with justification

File: src/mavlink-router/endpoint.h [hunk 2]
```diff
@@ -131,6 +134,9 @@ class Endpoint : public Pollable {
     const bool _crc_check_enabled;
     uint32_t _incomplete_msgs = 0;
     std::vector<uint16_t> _sys_comp_ids;
+
+private:
+    std::set<uint32_t> message_filter;
```
Comment (lucasdemarchi) [inline src/mavlink-router/endpoint.h]: I'm not a big fan of std::set. I'd rather use std::vector here. When we are filtering we will probably care for very few messages, so std::vector should even be faster for lookups.

File: src/mavlink-router/main.cpp [hunk 3]
```diff
@@ -224,19 +224,24 @@ static int add_endpoint_address(const char *name, size_t name_len, const char *i
...
-        free(conf->address);
         conf->address = strdup(ip);
         if (!conf->address) {
             ret = -ENOMEM;
             goto fail;
         }
-    }
-
-    if (!conf->address) {
+    } else {
```
Comment (lucasdemarchi) [inline src/mavlink-router/main.cpp]: all these changes in this function seem good, but should be in a separate commit as a cleanup. I think this function got reworked at some time and these checks remained, but don't make sense anymore.

Comment (lucasdemarchi) [reply src/mavlink-router/main.cpp]: In this one, btw: `ip` is mandatory... so just remove the `if (ip)` and let the check for ENOMEM. It would be good to put a `assert_or_return(ip != nullptr, -EINVAL)` in the beginning of the function.

File: src/mavlink-router/ulog.cpp [hunk 4]
```diff
@@ -191,7 +191,7 @@ int ULog::write_msg(const struct buffer *buffer)
         mavlink_msg_logging_ack_encode(LOG_ENDPOINT_SYSTEM_ID, MAV_COMP_ID_ALL, &msg, &ack);
         _send_msg(&msg, _target_system_id);
         /* message will be handled by MAVLINK_MSG_ID_LOGGING_DATA case */
-        [[gnu::fallthrough]];
+        /* fall through */
```
Comment (lucasdemarchi) [inline src/mavlink-router/ulog.cpp]: ?

File: src/mavlink-router/main.cpp [hunk 5]
```diff
@@ -667,11 +672,13 @@ static int parse_confs(ConfFile &conf)
         char *addr;
         bool eavesdropping;
         unsigned long port;
+        char *filter;
     };
     static const ConfFile::OptionsTable option_table_udp[] = {
         {"address", true,   ConfFile::parse_str_dup,    OPTIONS_TABLE_STRUCT_FIELD(option_udp, addr)},
         {"mode",    true,   parse_mode,                 OPTIONS_TABLE_STRUCT_FIELD(option_udp, eavesdropping)},
         {"port",    false,  ConfFile::parse_ul,         OPTIONS_TABLE_STRUCT_FIELD(option_udp, port)},
+        {"filter",  false,  ConfFile::parse_str_dup,    OPTIONS_TABLE_STRUCT_FIELD(option_udp, filter)},
```
Comment (lucasdemarchi) [inline src/mavlink-router/main.cpp]: I guess it would be useful to have a `ConfigFile::parse_ul_vector` and make it do the parsing. Then you just std::move<> the parsed vector instead of strdup() the string.

Comment (lucasdemarchi) [reply src/mavlink-router/main.cpp]: this can be left for later though, as it's more intrusive.

File: src/mavlink-router/endpoint.h (hunk 2 above)
Comment (MatejFranceskin) [reply src/mavlink-router/endpoint.h]: Replaced std::set with std::vector.

File: src/mavlink-router/ulog.cpp (hunk 4 above)
Comment (MatejFranceskin) [reply src/mavlink-router/ulog.cpp]: This removes compile warning, source:
https://stackoverflow.com/questions/45349079/how-to-use-attribute-fallthrough-correctly-in-gcc

File: src/mavlink-router/endpoint.cpp (hunk 1 above)
Comment (MatejFranceskin) [reply src/mavlink-router/endpoint.cpp]: Reverted - will create a new cleanup PR with justification.

File: src/mavlink-router/main.cpp (hunk 3 above)
Comment (MatejFranceskin) [reply src/mavlink-router/main.cpp]: Reverted - will create a new cleanup PR.

File: src/mavlink-router/main.cpp (hunk 5 above)
Comment (MatejFranceskin) [reply src/mavlink-router/main.cpp]: If you think that it can stay this way then I will not change it for now.

**Comment by MatejFranceskin** [comment]
Thank you for your review. Please take a look at the changes that I made to address the issues.

**Comment by dk7xe** [comment]
Thanx Matej! That really helps! Gerald
